bayes_kit/rhat.py

```diff
@@ -1,11 +1,88 @@
 import numpy as np
 from numpy.typing import NDArray, ArrayLike
+import scipy as sp
 
 FloatType = np.float64
 VectorType = NDArray[FloatType]
 SeqType = ArrayLike
 
 
+def split_chains(chains: list[SeqType]) -> list[SeqType]:
+    """
+    Return a list of the input chains split in half.  The result will be
+    a list twice as long as the input.  For example,
+    ```
+    >>> split_chains([[1, 2, 3, 4], [5, 6, 7]])
+    [[1, 2], [3, 4], [5, 6], [7]]
+    ```
+
+    Parameters:
+    chains: List of univariate Markov chains.
+
+    Returns:
+    List of input chains split in half.
+    """
+    return [list(arr) for chain in chains for arr in np.array_split(chain, 2)]
+
+
+def rank_chains(chains: list[SeqType]) -> list[SeqType]:
+    """
+    Returns a copy of the included Markov chains transformed with
+    ranks normalized to [0, 1] and an offset inverse CDF. The ranks
+    are ascending and start with 1 for the smallest value.
+
+    Parameters:
+    chains: list of univariate Markov chains
+
+    Returns:
+    List of chains with values replaced by transformed ranks.
+    """
+    if len(chains) == 0:
+        return chains
+    flattened = np.concatenate(chains)
+    ranks = flattened.argsort().argsort() + 1
+    reshaped_arrays = []
+    current_index = 0
+    for array in chains:
+        size = len(array)
+        reshaped_arrays.append(ranks[current_index : current_index + size])
+        current_index += size
+    return reshaped_arrays
+
+
+def rank_normalize_chains(chains: list[SeqType]) -> list[SeqType]:
+    """Return the rank-normalized version of the input chains.
+
+    The rank-normalized value for element `j` of list `i` is
+    ```
+    inverse_Phi((rank[i][j] - 3/8) / (size(chains) - 1/4),
```

Comment:
I'd argue the doc string is so simple that the code looks just like it.    I added a citation to the original 1958 paper, which is where the constants come from and a short note as to what the constants are for.  The other functions already have citations.

Reply:
Adding the reference is good, but I would suggest--first, jumping out of the context of this code to an outside reference can be a pretty distracting or train-of-thought-breaking thing when you just want to understand the  code (to the point that many readers may not even attempt it); and second, the paper is likely to provide an explanation that's in a register that may be difficult to parse for the casual reader (by which I guess I mean, for anybody who didn't immediately understand what was happening from the code itself). I'd really be looking just for a bit more intuition here, like explaining what rank-normalization is ("projecting ranks so that they're evenly distributed on a normal curve", if that's correct, etc).

Reply:
When C++ documents cosine, they don't write an intro to trig in the doc.  It says, "Computes the cosine of num (measured in radians)."  It assumes if you want to use trig functions that you know trig.  I would like to make roughly the same kind of decision with our doc.  

Inverse cdfs are standard, sample means are standard, and sample variance is standard, in very much the same way as cosine is standard.

The rank normalization isn't so standard that everyone knows it, but it uses a very simple function to map to (0, 1) evenly, then it uses the well-known normal inverse CDF to project back to the real line.  That's all there is to it.  If the reader doesn't know how inverse CDFs work, I don't think the doc to the `rank_normalize_chains` is the place for a tutorial.  

I wasn't sure what you mean by "evenly distributed on a normal curve" but the spacing will not be even when you project back through the inverse CDF.  That is, `inv_Phi([0.1, 0.2, ..., 0.9])` does not have even spaced values even though `[0.1, ..., 0.9]` does.  Of course, calling `inv_Phi` (a standard function) `sp.stats.norm.ppf` isn't helping anyone.


```python
>>> y = (np.arange(10) + 0.5) / 10

>>> y
array([0.05, 0.15, 0.25, 0.35, 0.45, 0.55, 0.65, 0.75, 0.85, 0.95])

>>> sp.stats.norm.ppf(y)
array([-1.64485363, -1.03643339, -0.67448975, -0.38532047, -0.12566135,
        0.12566135,  0.38532047,  0.67448975,  1.03643339,  1.64485363])
```

Reply:
P.S.  3/8 and 1/4 is to make sure the results are in (0, 1).  The exact values don't matter so much other than that you don't want to get too close to 0 or 1 or that pushes you close to -inf, +inf.   3/8 and 1/4 are just a reasonable compromise in keeping the values reasonable for various sizes of inputs.

Reply:
I don't think there's a need to write an intro to stats in the doc here; but "rank normalization" doesn't even have a Wikipedia entry & it's not "normalization" in the most obvious sense (of converting the values to a 0-1 range).

I think I'd be looking for something like "Rank normalization maps the ranks to the range (0, 1) and then returns the quantiles of the standard-normal distribution for the resulting values. (A small margin is first applied to avoid infinite values from the ppf function). \n\n Specifically, the rank-normalized value ..." This is pretty much the information that's present in the doc for the `rank_normalized_rhat` function, but when you're looking at this function, you probably don't realize you ought to look over there for it.

Reply:
I just copied in what you suggested (but without the parens).   So as not to have redundant doc, I then erased where I said something similar two paragraphs later.  

The definition of "normalization" being used here is in the general sense of

https://en.wikipedia.org/wiki/Normalization_(statistics)

and is very close to the specific sense of

https://en.wikipedia.org/wiki/Quantile_normalization

with a standard Gaussian distribution and the specific offsets introduced in the cited paper.